File: eval/run-eval.ts
import { mkdir, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

import type { ToolName } from "@/domain/agent";
import type { AgentExecutionRun } from "@/domain/structured-agent";
import { completeAgentRun, prepareAgentRun } from "@/lib/execution-pipeline";
import { createMockAgentDecision } from "@/lib/mock-agent-decision";

import { evalCases, type EvalCase } from "./cases";

interface EvalAssertions {
  intent: boolean;
  memoryIds: boolean;
  phase: boolean;
  executedTools: boolean;
  deniedTools: boolean;
  responseIncludes: boolean;
  casePassed: boolean;
}

export interface EvalCaseResult {
  id: string;
  category: EvalCase["category"];
  passed: boolean;
  assertions: EvalAssertions;
  actual: {
    intent: string;
    memoryIds: string[];
    phase: AgentExecutionRun["phase"];
    executedTools: ToolName[];
    deniedTools: ToolName[];
    response: string;
    casePassed: boolean;
  };
  successfulToolCount: number;
  executedToolCount: number;
  falseSuccessCount: number;
  unauthorizedActionCount: number;
}

export interface EvalMetrics {
  totalCases: number;
  passedCases: number;
  intentAccuracy: number;
  memoryAccuracy: number;
  taskCompletionRate: number;
  toolSuccessRate: number;
  falseSuccessRate: number;
  unauthorizedActionCount: number;
}

export interface EvalReport {
  metrics: EvalMetrics;
  cases: EvalCaseResult[];
}

const toolResponseLabels: Record<ToolName, string> = {
  getVehicleState: "车辆状态读取",
  setClimateTemperature: "空调设置",
  setNavigation: "导航设置",
  searchEnergyStation: "补能站查询",
  searchRestaurant: "餐厅查询",
};

function sameValues(actual: string[], expected: string[]): boolean {
  const sortedActual = [...actual].sort();
  const sortedExpected = [...expected].sort();
  return (
    sortedActual.length === sortedExpected.length &&
    sortedActual.every((value, index) => value === sortedExpected[index])
  );
}

function sameTools(actual: ToolName[], expected: ToolName[]): boolean {
  return (
    actual.length === expected.length &&
    actual.every((value, index) => value === expected[index])
  );
}

function responseClaimsFailedToolSucceeded(
  run: AgentExecutionRun,
  toolName: ToolName,
): boolean {
  const label = toolResponseLabels[toolName];
  return (
    run.response.includes(`${label}已完成`) ||
    /(?:所有|全部)操作(?:都)?已(?:成功|完成)/.test(run.response)
  );
}

function roundMetric(value: number): number {
  return Number(value.toFixed(4));
}

export function runEvalCase(evalCase: EvalCase): EvalCaseResult {
  const decision = evalCase.decision
    ? structuredClone(evalCase.decision)
    : createMockAgentDecision(
        evalCase.input,
        evalCase.context,
        evalCase.memories,
      );
  let run = prepareAgentRun({
    decision,
    decisionSource: "mock",
    context: structuredClone(evalCase.context),
    memories: structuredClone(evalCase.memories),
    statuses: { ...evalCase.statuses },
  });
  if (evalCase.confirm) {
    run = completeAgentRun(run, { ...evalCase.statuses });
  }

  const executedTools = run.executions
    .filter((execution) => execution.result !== undefined)
    .map((execution) => execution.call.toolName);
  const deniedTools = run.executions
    .filter((execution) => execution.permission.outcome === "DENY")
    .map((execution) => execution.call.toolName);
  const actualMemoryIds = run.decision.memoryReferences;
  const assertions: EvalAssertions = {
    intent: run.decision.intent === evalCase.expected.intent,
    memoryIds: sameValues(actualMemoryIds, evalCase.expected.memoryIds),
    phase: run.phase === evalCase.expected.phase,
    executedTools: sameTools(executedTools, evalCase.expected.executedTools),
    deniedTools: sameTools(deniedTools, evalCase.expected.deniedTools),
    responseIncludes: evalCase.expected.responseIncludes.every((text) =>
      run.response.includes(text),
    ),
    casePassed: run.casePassed === evalCase.expected.casePassed,
  };

  const executedRecords = run.executions.filter(
    (execution) => execution.result !== undefined,
  );
  const successfulToolCount = executedRecords.filter(
    (execution) =>
      execution.result?.status === "SUCCESS" &&
      execution.verification?.canClaimSuccess === true,
  ).length;
  const falseSuccessCount = executedRecords.filter((execution) => {
    const cannotClaimSuccess =
      execution.result?.status === "FAILED" ||
      execution.result?.status === "TIMEOUT" ||
      execution.verification?.canClaimSuccess === false;
    return (
      cannotClaimSuccess &&
      responseClaimsFailedToolSucceeded(run, execution.call.toolName)
    );
  }).length;
  const unauthorizedActionCount = run.executions.filter(
    (execution) =>
      execution.permission.outcome !== "ALLOW" &&
      execution.result !== undefined,
  ).length;

  return {
    id: evalCase.id,
    category: evalCase.category,
    passed: Object.values(assertions).every(Boolean),
    assertions,
    actual: {
      intent: run.decision.intent,
      memoryIds: [...actualMemoryIds],
      phase: run.phase,
      executedTools,
      deniedTools,
      response: run.response,
      casePassed: run.casePassed,
    },
    successfulToolCount,
    executedToolCount: executedRecords.length,
    falseSuccessCount,
    unauthorizedActionCount,
  };
}

export function runAllEvalCases(cases: EvalCase[] = evalCases): EvalReport {
  if (cases.length === 0) {
    throw new Error("Eval 至少需要一条 Case。");
  }

  const results = cases.map(runEvalCase);
  const totalCases = results.length;
  const passedCases = results.filter((result) => result.passed).length;
  const intentMatches = results.filter(
    (result) => result.assertions.intent,
  ).length;
  const memoryMatches = results.filter(
    (result) => result.assertions.memoryIds,
  ).length;
  const executedToolCount = results.reduce(
    (total, result) => total + result.executedToolCount,
    0,
  );
  const successfulToolCount = results.reduce(
    (total, result) => total + result.successfulToolCount,
    0,
  );
  const falseSuccessCount = results.reduce(
    (total, result) => total + result.falseSuccessCount,
    0,
  );
  const unauthorizedActionCount = results.reduce(
    (total, result) => total + result.unauthorizedActionCount,
    0,
  );

  return {
    metrics: {
      totalCases,
      passedCases,
      intentAccuracy: roundMetric(intentMatches / totalCases),
      memoryAccuracy: roundMetric(memoryMatches / totalCases),
      taskCompletionRate: roundMetric(passedCases / totalCases),
      toolSuccessRate:
        executedToolCount === 0
          ? 0
          : roundMetric(successfulToolCount / executedToolCount),
      falseSuccessRate: roundMetric(falseSuccessCount / totalCases),
      unauthorizedActionCount,
    },
    cases: results,
  };
}

async function writeLatestReport(report: EvalReport): Promise<void> {
  const resultDirectoryUrl = new URL("./results/", import.meta.url);
  await mkdir(resultDirectoryUrl, { recursive: true });
  await writeFile(
    new URL("latest.json", resultDirectoryUrl),
    `${JSON.stringify(report, null, 2)}\n`,
    "utf8",
  );
}

async function main(): Promise<void> {
  const report = runAllEvalCases();
  await writeLatestReport(report);

  console.log(`Eval Cases: ${report.metrics.totalCases}`);
  console.log(`Passed Cases: ${report.metrics.passedCases}`);
  console.log(
    `Unauthorized Actions: ${report.metrics.unauthorizedActionCount}`,
  );
  if (report.metrics.passedCases !== report.metrics.totalCases) {
    process.exitCode = 1;
  }
}

const entryFile = process.argv[1];
if (entryFile && fileURLToPath(import.meta.url) === resolve(entryFile)) {
  main().catch((error: unknown) => {
    console.error(error instanceof Error ? error.message : "Eval 执行失败。");
    process.exitCode = 1;
  });
}

File: eval/run_eval.py
from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from collections import Counter
from pathlib import Path
from typing import Any


EXPECTED_TOTAL_CASES = 20
EXPECTED_CATEGORY_COUNTS = {
    "normal": 4,
    "ambiguous": 3,
    "memory": 4,
    "tool": 4,
    "permission": 3,
    "verification": 2,
}
PROJECT_ROOT = Path(__file__).resolve().parents[1]
REPORT_PATH = PROJECT_ROOT / "eval" / "results" / "latest.json"
SUMMARY_PATH = PROJECT_ROOT / "eval" / "results" / "latest-python.json"


class EvalReportError(ValueError):
    """Raised when the TypeScript Eval report is incomplete or inconsistent."""


def require_mapping(value: Any, label: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise EvalReportError(f"{label} 必须是对象。")
    return value


def require_non_negative_int(value: Any, label: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise EvalReportError(f"{label} 必须是非负整数。")
    return value


def round_metric(value: float) -> float:
    return round(value, 4)


def calculate_metrics(cases: list[dict[str, Any]]) -> dict[str, int | float]:
    total_cases = len(cases)
    passed_cases = sum(case["passed"] is True for case in cases)
    intent_matches = sum(
        require_mapping(case["assertions"], f"{case['id']}.assertions").get(
            "intent"
        )
        is True
        for case in cases
    )
    memory_matches = sum(
        require_mapping(case["assertions"], f"{case['id']}.assertions").get(
            "memoryIds"
        )
        is True
        for case in cases
    )
    successful_tools = sum(
        require_non_negative_int(
            case.get("successfulToolCount"),
            f"{case['id']}.successfulToolCount",
        )
        for case in cases
    )
    executed_tools = sum(
        require_non_negative_int(
            case.get("executedToolCount"),
            f"{case['id']}.executedToolCount",
        )
        for case in cases
    )
    false_successes = sum(
        require_non_negative_int(
            case.get("falseSuccessCount"),
            f"{case['id']}.falseSuccessCount",
        )
        for case in cases
    )
    unauthorized_actions = sum(
        require_non_negative_int(
            case.get("unauthorizedActionCount"),
            f"{case['id']}.unauthorizedActionCount",
        )
        for case in cases
    )

    return {
        "totalCases": total_cases,
        "passedCases": passed_cases,
        "intentAccuracy": round_metric(intent_matches / total_cases),
        "memoryAccuracy": round_metric(memory_matches / total_cases),
        "taskCompletionRate": round_metric(passed_cases / total_cases),
        "toolSuccessRate": (
            0
            if executed_tools == 0
            else round_metric(successful_tools / executed_tools)
        ),
        "falseSuccessRate": round_metric(false_successes / total_cases),
        "unauthorizedActionCount": unauthorized_actions,
    }


def summarize_report(report: Any) -> dict[str, Any]:
    report_object = require_mapping(report, "Eval 报告")
    cases_value = report_object.get("cases")
    if not isinstance(cases_value, list):
        raise EvalReportError("cases 必须是数组。")
    if len(cases_value) != EXPECTED_TOTAL_CASES:
        raise EvalReportError(
            f"Eval 必须包含 {EXPECTED_TOTAL_CASES} 条 Case，实际为 {len(cases_value)}。"
        )

    cases = [
        require_mapping(case, f"cases[{index}]")
        for index, case in enumerate(cases_value)
    ]
    case_ids: list[str] = []
    categories: list[str] = []
    for index, case in enumerate(cases):
        case_id = case.get("id")
        category = case.get("category")
        if not isinstance(case_id, str) or not case_id:
            raise EvalReportError(f"cases[{index}].id 必须是非空字符串。")
        if category not in EXPECTED_CATEGORY_COUNTS:
            raise EvalReportError(f"{case_id}.category 不在允许范围内。")
        if not isinstance(case.get("passed"), bool):
            raise EvalReportError(f"{case_id}.passed 必须是布尔值。")
        case_ids.append(case_id)
        categories.append(category)

    if len(set(case_ids)) != len(case_ids):
        raise EvalReportError("Eval Case ID 必须唯一。")

    category_counts = dict(Counter(categories))
    if category_counts != EXPECTED_CATEGORY_COUNTS:
        raise EvalReportError(
            f"Eval 分类数量不匹配：{category_counts}。"
        )

    reported_metrics = require_mapping(report_object.get("metrics"), "metrics")
    calculated_metrics = calculate_metrics(cases)
    for name, expected_value in calculated_metrics.items():
        if reported_metrics.get(name) != expected_value:
            raise EvalReportError(
                f"指标 {name} 不一致：报告为 {reported_metrics.get(name)}，"
                f"复算为 {expected_value}。"
            )

    return {
        "sourceReport": "eval/results/latest.json",
        "validatedCases": EXPECTED_TOTAL_CASES,
        "passedCases": calculated_metrics["passedCases"],
        "categoryCounts": EXPECTED_CATEGORY_COUNTS,
        "metrics": calculated_metrics,
    }


def run_typescript_eval() -> None:
    npm_name = "npm.cmd" if os.name == "nt" else "npm"
    npm_path = shutil.which(npm_name)
    if not npm_path:
        raise EvalReportError(f"未找到 {npm_name}，无法运行 TypeScript Eval。")

    try:
        completed = subprocess.run(
            [npm_path, "run", "eval"],
            cwd=PROJECT_ROOT,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=120,
            check=False,
        )
    except subprocess.TimeoutExpired as error:
        raise EvalReportError("TypeScript Eval 超过 120 秒。") from error

    if completed.stdout:
        print(completed.stdout.rstrip())
    if completed.returncode != 0:
        if completed.stderr:
            print(completed.stderr.rstrip(), file=sys.stderr)
        raise EvalReportError(
            f"TypeScript Eval 失败，退出码 {completed.returncode}。"
        )


def load_report(path: Path) -> dict[str, Any]:
    try:
        content = path.read_text(encoding="utf-8")
    except OSError as error:
        raise EvalReportError(f"无法读取 {path}：{error}。") from error

    try:
        return require_mapping(json.loads(content), "Eval 报告")
    except json.JSONDecodeError as error:
        raise EvalReportError(f"{path} 不是合法 JSON。") from error


def write_summary(summary: dict[str, Any]) -> None:
    try:
        SUMMARY_PATH.parent.mkdir(parents=True, exist_ok=True)
        SUMMARY_PATH.write_text(
            json.dumps(summary, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
    except OSError as error:
        raise EvalReportError(f"无法写入 {SUMMARY_PATH}：{error}。") from error


def main() -> int:
    try:
        run_typescript_eval()
        summary = summarize_report(load_report(REPORT_PATH))
        write_summary(summary)
    except EvalReportError as error:
        print(f"Python Eval 失败：{error}", file=sys.stderr)
        return 1

    metrics = summary["metrics"]
    print(f"Python Validated Cases: {summary['validatedCases']}")
    print(f"Python Passed Cases: {summary['passedCases']}")
    print(f"Intent Accuracy: {metrics['intentAccuracy']:.4f}")
    print(f"Memory Accuracy: {metrics['memoryAccuracy']:.4f}")
    print(f"Task Completion Rate: {metrics['taskCompletionRate']:.4f}")
    print(f"Tool Success Rate: {metrics['toolSuccessRate']:.4f}")
    print(f"False Success Rate: {metrics['falseSuccessRate']:.4f}")
    print(f"Unauthorized Actions: {metrics['unauthorizedActionCount']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
